cache/network/upload.go

```diff
@@ -32,39 +70,295 @@ func (u DefaultUploader) Upload(ctx context.Context, params UploadParams, logger
+	logger.Debugf("Complete multipart upload")
+	response, err := client.completeMultipartUpload(multipartResp.ID, etags)
 	if err != nil {
-		return fmt.Errorf("failed to finalize upload: %w", err)
+		return fmt.Errorf("failed to complete multipart upload: %w", err)
 	}
 
-	logger.Debugf("Upload acknowledged")
+	logger.Debugf("Multipart upload completed")
 	logResponseMessage(response, logger)
 
 	return nil
 }
 
+func (u DefaultUploader) uploadChunks(ctx context.Context, archivePath string, response prepareMultipartUploadResponse, logger log.Logger) ([]string, error) {
+	file, err := os.Open(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open archive file: %w", err)
+	}
+	defer func() {
+		if err := file.Close(); err != nil {
+			logger.Errorf("close archive: %v", err)
+		}
+	}()
+
+	chunkSize := response.ChunkSizeBytes
+	numChunks := len(response.URLs)
+
+	chunks := make([][]byte, numChunks)
+	currentOffset := int64(0)
+
+	for i := 0; i < numChunks; i++ {
+		currentChunkSize := chunkSize
+		if i == numChunks-1 {
+			currentChunkSize = response.LastChunkSizeBytes
+		}
+
+		// seek to correct position
+		_, err := file.Seek(currentOffset, io.SeekStart)
+		if err != nil {
+			return nil, fmt.Errorf("failed to seek to position %d for chunk %d: %w", currentOffset, i+1, err)
+		}
+
+		// read chunk from file
+		chunk := make([]byte, currentChunkSize)
+		n, err := io.ReadFull(file, chunk)
+		if err != nil && err != io.ErrUnexpectedEOF {
+			return nil, fmt.Errorf("failed to read chunk %d: %w", i+1, err)
+		}
+
+		if n == 0 {
+			return nil, fmt.Errorf("unexpected end of file at chunk %d", i+1)
+		}
+
+		chunks[i] = chunk[:n]
+		currentOffset += int64(n)
+	}
+
+	// MaxRetryPerChunk is controls how many times to interrupt and retry a slow chunk upload.
+	// If zero, the chunk download monitoring is disabled and the chunk download won't be interrupted.
+	maxRetryPerChunk := 3
+
+	// ChunkRetryThreshold is the deviation from the moving average (of chunks uploaded so far) after which a chunk is interrupted and retried.
+	chunkRetryThreshold := 10 * time.Second
+
+	var stats chunkStatistics
+
+	type chunkResult struct {
+		index int
+		etag  string
+		err   error
+	}
+
+	resultChan := make(chan chunkResult, numChunks)
+	semaphore := make(chan struct{}, getDefaultConcurrency())
+
+	for i, uploadURL := range response.URLs {
+		go func(index int, url prepareMultipartUploadURL, chunkData []byte) {
+			semaphore <- struct{}{}
+			defer func() { <-semaphore }()
+
+			var etag string
+			var uploadErr error
+
+			for attempt := 0; attempt < maxRetryPerChunk; attempt++ {
+				logger.Debugf("Uploading chunk %d/%d (attempt %d/%d) [finished=%d] [avg=%v]",
+					index+1, numChunks, attempt+1, maxRetryPerChunk,
+					stats.getFinishedCount(), stats.average().Round(time.Second))
+
+				start := time.Now()
+
+				chunkCtx, cancelChunk := context.WithCancel(ctx)
+
+				if attempt < maxRetryPerChunk-1 {
+					go func() {
```

Comment:
This can also be extracted btw, let's aim to have each func layer separate

To share the context btw, you can use a struct that holds the "context", i.e. the results chan and the stats and the params ofc